entity-service-change-event-generator/src/main/java/org/hypertrace/entity/service/change/event/impl/EntityChangeEventGeneratorImpl.java

```diff
@@ -147,6 +172,67 @@ private void sendUpdateNotification(
     }
   }
 
+  private boolean shouldSendNotification(Entity prevEntity, Entity currEntity) {
+    boolean isAttributesChanged = false, isCoreEntityChanged = false;
+    for (FieldDescriptor fieldDescriptor : Entity.getDescriptor().getFields()) {
+      switch (fieldDescriptor.getNumber()) {
+        case Entity.ATTRIBUTES_FIELD_NUMBER:
+          isAttributesChanged = checkIfAttributesChanged(prevEntity, currEntity);
+          break;
+        default:
+          boolean isFieldChanged =
+              !prevEntity.getField(fieldDescriptor).equals(currEntity.getField(fieldDescriptor));
+          if (!isCoreEntityChanged) {
+            isCoreEntityChanged = isFieldChanged;
+          }
+          break;
+      }
+    }
+
+    return isCoreEntityChanged || isAttributesChanged;
+  }
+
+  private boolean checkIfAttributesChanged(Entity prevEntity, Entity currEntity) {
+    MapDifference<String, AttributeValue> attributesDiff =
+        Maps.difference(prevEntity.getAttributesMap(), currEntity.getAttributesMap());
+    String entityType = prevEntity.getEntityType();
+
+    List<Entry<String, AttributeValue>> addedAttributes =
+        attributesDiff.entriesOnlyOnRight().entrySet().stream()
+            .filter(attributeKey -> !shouldSkipAttribute(entityType, attributeKey.getKey()))
+            .collect(Collectors.toList());
+
+    List<Entry<String, AttributeValue>> deletedAttributes =
+        attributesDiff.entriesOnlyOnLeft().entrySet().stream()
+            .filter(
+                attributeKey ->
+                    !(this.entityToAttributeChangeSkipMap
+                        .get(entityType)
+                        .contains(attributeKey.getKey())))
+            .collect(Collectors.toList());
+
+    List<Entry<String, ValueDifference<AttributeValue>>> updatedAttributes =
+        attributesDiff.entriesDiffering().entrySet().stream()
+            .filter(
+                attributeKey ->
+                    !(this.entityToAttributeChangeSkipMap
+                        .get(entityType)
+                        .contains(attributeKey.getKey())))
+            .collect(Collectors.toList());
+
+    return addedAttributes.size() > 0
+        || deletedAttributes.size() > 0
+        || updatedAttributes.size() > 0;
+  }
+
+  private boolean isCoreEntityChanged(Entity prevEntity, Entity currEntity) {
```

Comment:
What's a core entity? And do we even support changing an entity name (even if we do, the name should also be an attribute)

Reply:
My assumption was first class params can also change. we should skip those cases too. Can I assume at least for entity cases that it will always be attributes that might change? Because of this only i have to play with proto fields descriptor that you mentioned above

Reply:
I would just assume first class params can not change. That would open up a can of worms in terms of how to specify it etc. anyway, and since we shouldn't be using first class params in the first place (at least, anything written in the last 2 or so years), we don't need to implement functionality for it.